compiler/rustc_type_ir/src/fold.rs

```diff
@@ -46,6 +46,12 @@
 //! ```
 use crate::{visit::TypeVisitable, Interner};
 
+#[cfg(feature = "nightly")]
+type Never = !;
+
+#[cfg(not(feature = "nightly"))]
+type Never = std::convert::Infallible;
```

Comment:
Seems to have problems with type inference later on in the crate:'
```
error[E0284]: type annotations needed
   --> compiler/rustc_type_ir/src/fold.rs:223:18
    |
223 |     type Error = Never;
    |                  ^^^^^ cannot infer type
    |
    = note: cannot satisfy `<F as FallibleTypeFolder<I>>::Error == _`
```

Reply:
Huh 🤔

Reply:
So the MRE is this: [(play)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cb498d3587f8d0983734acbfe50005fc).

`FallibleTypeFolder<Error = Never>` from `TypeFolder` super-traits gets checked against `type Error = Never` in the blanket impl of `FallibleTypeFolder` for `F: TypeFolder`.

First of all, are cycles like this even allowed? I.e. can you satisfy a super-trait bound by a blanket impl that asks for the same trait? I would think that this causes a cycle...

Second of all, can't we just remove the super trait? It seems like all the things that can be proven with the supertrait bound can be proven via the blanket impl (that might be wrong, not sure).

Lastly, why the compiler can't evaluate this projection here? It seems to be fine in [other places](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=307c44761f5d3c71604c00dcf87d7e0d), and I don't see what can't it infer...

Reply:
I don't see why it should cause a cycle... impls (blanket or otherwise) can be provided even where supertraits are not satisfied.

We can't "just" remove the supertrait because the default implementations of the subtrait's provided methods ultimately call into the supertrait — and I don't think the existence of the blanket impl is provable there, within the trait definition.

Reply:
Even more absurd: [(play)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9ea422a172a4945262f2e98698596a94).

Reply:
Okay, so the error arises whenever the associated type doesn't match the supertrait constraint: [(play)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a822ebe2a543ce1ea4a03302879841b6).

Not sure why the compiler is unable to prove that the `Never` types are equal, though...